ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh

```diff
@@ -445,31 +449,82 @@ EOF
+# $2: Either "workers" or "masters" to enable spot instances on the
+#     compute or control machines, respectively.
 function inject_spot_instance_config() {
   local dir=${1}
+  local mtype=${2}
 
-  if [ ! -f /tmp/yq ]; then
-    curl -L "https://github.com/mikefarah/yq/releases/download/3.3.0/yq_linux_$( get_arch )" \
-    -o /tmp/yq && chmod +x /tmp/yq
-  fi
 
-  PATCH="${SHARED_DIR}/machinesets-spot-instances.yaml.patch"
-  cat > "${PATCH}" << EOF
-spec:
-  template:
-    spec:
-      providerSpec:
-        value:
-          spotMarketOptions: {}
-EOF
+  get_yq
 
-  for MACHINESET in $dir/openshift/99_openshift-cluster-api_worker-machineset-*.yaml; do
-    /tmp/yq m -x -i "${MACHINESET}" "${PATCH}"
-    echo "Patched spotMarketOptions into ${MACHINESET}"
+  # Find manifest files
+  local manifests=
+  case "${mtype}" in
+    masters)
+      # Spot masters only works for CAPA today. Validate.
+      # TODO: Add support for tf masters after https://github.com/openshift/installer/pull/8349 lands.
+      #       Add validation so we don't end up silently ignoring SPOT_MASTERS requests.
+      if ! [[ "$(/tmp/yq r "${SHARED_DIR}/install-config.yaml" 'featureGates[*]')" =~ ClusterAPIInstall(AWS)?=true ]]; then
+        echo "ERROR: Spot instances for masters can only be used with CAPI installs."
+        exit 1
+      fi
+      manifests="${dir}/cluster-api/machines/10_inframachine_*.yaml \
+        ${dir}/openshift/99_openshift-machine-api_master-control-plane-machine-set.yaml \
+        ${dir}/openshift/99_openshift-cluster-api_*-machines-*.yaml"
+      ;;
+    workers)
+      manifests="${dir}/openshift/99_openshift-cluster-api_*-machineset-*.yaml"
+      ;;
+    *)
+      echo "ERROR: Invalid machine type '$mtype' passed to inject_spot_instance_config; expected 'masters' or 'workers'"
+      exit 1
+      ;;
+  esac
+
+  # Inject spotMarketOptions into the appropriate manifests
+  local prefix=
+  local found=false
+  # Don't rely on file names; iterate through all the manifests and match
+  # by kind.
+  for manifest in $manifests; do
+    kind=$(/tmp/yq r "${manifest}" kind)
+    case "${kind}" in
+      MachineSet)  # Workers, both tf and CAPA, run through MachineSet today.
+          [[ "${mtype}" == "workers" ]] || continue
+          prefix='spec.template.spec.providerSpec.value'
+          ;;
+      AWSMachine)  # CAPA masters
+          [[ "${mtype}" == "masters" ]] || continue
+          prefix='spec'
+          ;;
+      Machine)  # tf masters during install
+          [[ "${mtype}" == "masters" ]] || continue
+          prefix='spec.providerSpec.value'
+          ;;
+      ControlPlaneMachineSet)  # masters reconciled after install
+          [[ "${mtype}" == "masters" ]] || continue
+          prefix='spec.template.machines_v1beta1_machine_openshift_io.spec.providerSpec.value'
+          ;;
+      *)
+          continue
+          ;;
+    esac
+    found=true
```

Comment:
There is no failure in the switch, wont this always be true?

Reply:
Thanks for the look!

The `continue`s (LL498, 502, 506, 510, 514) should skip this. The idea being to error in the case that `SPOT_$something` was requested, but didn't match anything (L525).

...but you got me questioning my bash chops -- does `continue` go all the way to the next iteration of `for` (as intended here) or does it just drop out of the `case`?? So I tested it with this little scriptlet:

```bash
#!/bin/bash

found=false
for x in "$@"; do
  case "$x" in
  foo) echo found ;;
  *) continue ;;
  esac
  found=true
done
echo "found? $found"
```

Result:

```
(.venv) [efried@efried 416]$ ./scr
found? false
(.venv) [efried@efried 416]$ ./scr blah
found? false
(.venv) [efried@efried 416]$ ./scr blah foo
found
found? true
```

So I think we're good :+1: